cmd/talosctl/cmd/talos/root.go

```diff
@@ -123,3 +127,92 @@ var Commands []*cobra.Command
 func addCommand(cmd *cobra.Command) {
 	Commands = append(Commands, cmd)
 }
+
+// completeResource represents tab complete options for `ls` and `ls *` commands.
+func completePathFromNode(inputPath string) []string {
+	pathToSearch := inputPath
+
+	// If the pathToSearch is empty, use root '/'
+	if pathToSearch == "" {
+		pathToSearch = "/"
+	}
+
+	var paths map[string]struct{}
+
+	// search up one level to find possible completions
+	if pathToSearch != "/" && !strings.HasSuffix(pathToSearch, "/") {
+		splits := strings.Split(pathToSearch, "/")
+		// we need a trailing slash to search for items in a directory
+		pathToSearch = strings.Join(splits[:len(splits)-1], "/") + "/"
+	}
+
+	paths = getPathFromNode(pathToSearch)
+
+	result := make([]string, 0, len(paths))
+
+	for k := range paths {
+		result = append(result, k)
```

Comment:
really a nit, but if we filter here, we could avoid building the list twice

or it might be even more fancy to avoid building full list in `getPathFromNode` by applying filtering there? this way if the directory listing is huge, we never have to store it all in memory

Reply:
fixed